### PR TITLE
[Cookies] Expire the AB test cookie in a year

### PR DESCRIPTION
**Pull request by zhhyu**
We should expire any cookie in 13 months, no matter whether it's in EU or not.
We upgrade the cookie to V3, in order to handle these already dropped cookies.

### CODE REVIEW

**Review by joelverhagen — APPROVED** [review]

**Comment by joelverhagen** [comment]
What about existing cookies on people's devices with an expiration that's too long?

**Comment by zhhyu** [comment]
> What about existing cookies on people's devices with an expiration that's too long?

Good question. I create a tracking issue for this: https://github.com/NuGet/Engineering/issues/3463
We should be able to reset the expiration time as one year by checking its expiration time.

**Comment by joelverhagen** [comment]
> We should be able to reset the expiration time as one year by checking its expiration time.

AFAIK expiration time is not sent to the server.

**Comment by zhhyu** [comment]
> > We should be able to reset the expiration time as one year by checking its expiration time.
> 
> AFAIK expiration time is not sent to the server.



> > We should be able to reset the expiration time as one year by checking its expiration time.
> 
> AFAIK expiration time is not sent to the server.

I have tested it, and I think you are right. We may need to confirm to see whether we are allowed to renew the cookie.

**Review by loic-sharma — APPROVED** [review]

File: src/NuGetGallery/Infrastructure/ABTestEnrollmentFactory.cs [hunk 1]
```diff
@@ -140,11 +143,45 @@ private bool TryDeserializeStateVer2(string serialized, out ABTestEnrollment enr
...
+
+                return true;
+            }
+            catch (JsonException)
+            {
+                return false;
+            }
+        }
+
+        private bool TryDeserializeStateVer3(string serialized, out ABTestEnrollment enrollment)
```
Comment (joelverhagen) [inline src/NuGetGallery/Infrastructure/ABTestEnrollmentFactory.cs:165]: Schema version 2 and 3 are the same exception the version property. Would it be possible to share more code? Even the  `StateVersion2` model could be renamed to `StateVersion2And3`.

**Review by joelverhagen — APPROVED** [review]
Try to share more code, but it's pretty much fine as-is.

File: src/NuGetGallery/Infrastructure/ABTestEnrollmentFactory.cs (hunk 1 above)
Comment (zhhyu) [reply src/NuGetGallery/Infrastructure/ABTestEnrollmentFactory.cs:165]: I have combined StateVersion2 and StateVersion3. I prefer to leave others as it is now. I thought about whether it's better to fully integrate version 2 and version3, but I feel that it's more readable and maintainable as the current one. Let me know if you have other thoughts here.

**Pull request merged**